### ISSUE TITLE
Some error handling in precertificate preIssuer AKI replacement code seems wrong

### ISSUE THREAD

**Issue by briansmith** [google/certificate-transparency-go#135]
See https://github.com/google/certificate-transparency-go/blob/4cf7f9eda5a5c4bd4b97acdc5935aa38afa660af/x509/x509.go#L1036-L1038:

```go
tbs.Extensions[keyAt].Value = issuerKeyID
if err != nil {
	return nil, fmt.Errorf("failed to marshal new auth key ID: %v", err)
}
```

It seems like the `err != nil` condition is dead code. Notice that `err` isn't set in any of the immediately preceding lines. I'm not sure what's supposed to be happening here.

**Comment by daviddrysdale** [comment]
Good spot -- I believe that's left over from an [earlier iteration](https://github.com/google/certificate-transparency-go/commit/2c2ae9004e05e551f3d592b629c1fe66af21af11) that [got replaced](https://github.com/google/certificate-transparency-go/commit/308855f505e444c931) with something that didn't need an `asn1.Marshal`.

**Comment by briansmith** [comment]
I noticed another problem with this error handling logic:

```go
if issuerKeyID == nil {
    return nil, fmt.Errorf("issuer has no auth-key-id extension")
}
```


```go
if keyAt < 0 {
	return nil, fmt.Errorf("pre-cert has no authority-key-id extension to update")
}
```

AKI isn't mandatory for end-entity certs (AFAICT from reading RFC 5280) nor is it mandatory for the pre-issuer certificate if the end-entity certificate didn't have an AKI, according to the CT spec. The check for the precert's AKI should be done first; only if it has an AKI should the pre-issuer certs' AKI be checked.

**Comment by daviddrysdale** [comment]
True, the code will currently only cope with a cert+preIssuer that both have AKIs; that should probably be more flexible.

By the way, do you know of systems that use the preIssuer mechanism?  I'm not sure if we've ever encountered it in the wild (and hence the overly-specific code here)...

**Comment by briansmith** [comment]
> By the way, do you know of systems that use the preIssuer mechanism? I'm not sure if we've ever encountered it in the wild (and hence the overly-specific code here)...

I don't. I only noticed these issues because some code I'd written recently hadn't considered the need for the AKI to be handled differently when the preIssuer mechanism is used.

**Comment by daviddrysdale** [comment]
Fixed in #137
